scripts/check_blocking_packages.py

```diff
@@ -0,0 +1,187 @@
+    if 'packages' in repo_data:
+        current_package_names = current_package_names | set(
+            repo_data['packages'])
+    else:
+        # Assume the repo name is the same as the package name
+        current_package_names.add(repo)
+
+# Construct a dictionary where keys are repository names and values are a list
+# of the missing dependencies for that repository
+# TODO: Could also check which packages are blocked by the input package
+# since DependencyWalker offers a get_depends_on() function
+
+blocked_repos = {}
+unblocked_repos = set()
+total_blocking_repos = set()
+
+for repository_name in repository_names:
+    # Check all the dependencies for the repository
+    if not repository_name in prev_distro_file.repositories:
+        continue
+
+    repo = prev_distro_file.repositories[repository_name]
+    # Check if the repo has a release, skip it if it doesn't
+    release_repo = repo.release_repository
+    if release_repo is None:
+        continue
+    if release_repo.version is None:
+        continue
+
+    package_dependencies = set()
+    packages = release_repo.package_names
+    # Accumulate all dependencies for those packages
+    for package in packages:
+        recursive_dependencies = dependency_walker.get_recursive_depends(
+            package, ['build', 'run', 'buildtool'], ros_packages_only=True,
+            limit_depth=dependency_walker_depth)
+        package_dependencies = package_dependencies.union(
+            recursive_dependencies)
+
+    # For all package dependencies, check if they are released yet
+    unreleased_pkgs = package_dependencies.difference(
+        current_package_names)
+    # remove the packages which this repo provides.
+    unreleased_pkgs = unreleased_pkgs.difference(packages)
+    # Now get the repositories for these packages.
+    blocking_repos = set(prev_distro_file.release_packages[pkg].repository_name
+                         for pkg in unreleased_pkgs)
+    if len(blocking_repos) == 0:
+        unblocked_repos.add(repository_name)
+    else:
+        # Get the repository for the unreleased packages
+        blocked_repos[repository_name] = blocking_repos
+        total_blocking_repos |= blocking_repos
+
+unblocked_blocking_repos = total_blocking_repos.intersection(unblocked_repos)
+unblocked_leaf_repos = unblocked_repos.difference(unblocked_blocking_repos)
+
+# Double-check repositories that we think are leaf repos
+for repo in unblocked_leaf_repos:
+    depends_on = dependency_walker.get_recursive_depends_on(
+        package, ['build', 'run', 'buildtool'])
+    if len(depends_on) != 0:
+        # There are packages that depend on this "leaf", but we didn't find
+        # them initially because they weren't related to our inputs
+        unblocked_blocking_repos.add(repo)
+unblocked_leaf_repos = unblocked_leaf_repos.difference(
+    unblocked_blocking_repos)
+
+if len(blocked_repos.keys()) > 0:
+    print("The following repos are blocked by other packages:")
```

Comment:
I think showing the block repositories would be more valuable rather then the blocking packages. It would also make the list shorter hence more readable.